scientific-computing-with-python/learn-inheritance-by-building-an-equation-solver/quadratic_equation_solver.py

```diff
@@ -0,0 +1,146 @@
+                    terms.insert(0, f'{coefficient:+}x')
+                else:
+                    terms.insert(0, f'{coefficient:+}x**{n}')
+        equation_string = ' '.join(terms)
+        equation_string += ' = 0'
+        return re.sub(r'(?<!\d)1(?=x)', '', equation_string.strip('+'))
+    
+    @property
+    @abstractmethod
+    def grade(self):
+        ...
+
+    @property
+    @abstractmethod
+    def type(self):
+        ...
+    
+    @abstractmethod
+    def solve(self):
+        ...
+
+    @abstractmethod
+    def analyze(self):
+        ...
+
+    
+class LinearEquation(Equation):
+    @property
+    def grade(self):
+        return 1
+    
+    @property
+    def type(self):
+        return 'Linear Equation'
+
+    def solve(self):
+        x = - self.coefficients[0] / self.coefficients[1]
+        self.results.append(f'x = {round(x, 3)}')
+        return [x]
+    
+    def analyze(self):
+        self.details.append(f'slope = {self.coefficients[1]:>16}\ny-intercept = {self.coefficients[0]:>10}')
+        return {'slope': self.coefficients[1], 'intercept': self.coefficients[0]}
+
+
+class QuadraticEquation(Equation):
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.delta = self.coefficients[1]**2 - 4 * self.coefficients[2] * self.coefficients[0]
+
+    @property
+    def grade(self):
+        return 2
+    
+    @property
+    def type(self):
+        return 'Quadratic Equation'
+
+    def solve(self):
+        if self.delta < 0:
+            self.results.append('No real roots')            
+            return []
+
+        x1 = (- self.coefficients[1] + (self.delta)**0.5)/(2 * self.coefficients[2])
+        x2 = (- self.coefficients[1] - (self.delta)**0.5)/(2 * self.coefficients[2])        
+        if self.delta == 0:
+            self.results.append(f'x = {round(x1, 3)}')            
+            return [x1]
+
+        self.results.extend([f'x1 = {round(x1, 3)}', f'x2 = {round(x2, 3)}'])        
+        return [x1, x2]
+
+    def analyze(self):
+        x = - self.coefficients[1] / (2 * self.coefficients[2])
+        y = self.coefficients[2] * x**2 + self.coefficients[1] * x + self.coefficients[0]
+        vertex = {'x': x, 'y': y}                    
+        if self.coefficients[2] > 0:
+            concavity = 'upwards'
+            vertex['m'] = 'min'            
+        else:
+            concavity = 'downwards'
+            vertex['m'] = 'max'
+        coord = f'{(round(x, 3), round(x, 3))}'
+        self.details.append(f'concavity = {concavity:>12}\n{vertex["m"]} = {coord:>18}')
+        return {'vertex': vertex, 'concavity': concavity}
+    
+
+class Solver:
+    def __init__(self, equation):
+        if not isinstance(equation, Equation):
+            raise TypeError("Argument must be an Equation object")
+        self.equation = equation
+
+    def _solve(self):
+        self.equation.solve()
+        self.equation.analyze()
+    
+    def create_output(self):
```

Comment:
```suggestion
    def print_output(self):
```
I would call this like this as it prints, but I think it's fine otherwise